src/provider/provider_tracking.c

```diff
@@ -429,6 +466,132 @@ static const char *trackingName(void *provider) {
+                //    the cache. Alternative approach could be insert empty
+                //    cache_value and only if insert succeed get actual IPC
+                //    handle and fill the cache_value structure under the lock.
+                //    But this case should be rare enough.
+                // 2. critnib failed to allocate memory internally. We need
+                //    to cleanup and return corresponding error.
+                free(cache_value);
+                ret = umfMemoryProviderPutIPCHandle(p->hUpstream, ipcData);
+                if (ret != UMF_RESULT_SUCCESS) {
+                    return ret;
+                }
+                if (insRes == ENOMEM) {
+                    return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+                }
+            }
+        }
+    } while (!cached);
+
+    return ret;
+}
+
+static umf_result_t trackingPutIpcHandle(void *provider, void *ipcData) {
+    (void)provider;
+    (void)ipcData;
+    // We just keep ipcData in the provider->ipcCache.
+    // The actual Put is called inside trackingFree
+    return UMF_RESULT_SUCCESS;
+}
+
+static size_t getDataSizeFromIpcHandle(const void *ipcData) {
+    // This is hack to get size of memory pointed by IPC handle.
+    // tracking memory provider gets only provider-specific data
+    // pointed by ipcData, but the size of allocation tracked
+    // by umf_ipc_data_t. We use this trick to get pointer to
+    // umf_ipc_data_t data because the ipcData is
+    // the Flexible Array Member of umf_ipc_data_t.
+    umf_ipc_data_t *ipcUmfData =
+        (umf_ipc_data_t *)((uint8_t *)ipcData - sizeof(umf_ipc_data_t));
+    return ipcUmfData->size;
+}
+
+static umf_result_t trackingOpenIpcHandle(void *provider, void *ipcData,
+                                          void **ptr) {
+    umf_tracking_memory_provider_t *p =
+        (umf_tracking_memory_provider_t *)provider;
+    umf_result_t ret = UMF_RESULT_SUCCESS;
+
+    ret = umfMemoryProviderOpenIPCHandle(p->hUpstream, ipcData, ptr);
+    if (ret != UMF_RESULT_SUCCESS) {
+        return ret;
+    }
+    size_t bufferSize = getDataSizeFromIpcHandle(ipcData);
+    ret = umfMemoryTrackerAdd(p->hTracker, p->pool, *ptr, bufferSize);
+    if (ret != UMF_RESULT_SUCCESS && p->hUpstream) {
+        if (umfMemoryProviderCloseIPCHandle(p->hUpstream, *ptr)) {
+            // TODO: LOG
+        }
+    }
+    return ret;
+}
+
+static umf_result_t trackingCloseIpcHandle(void *provider, void *ptr) {
+    umf_tracking_memory_provider_t *p =
+        (umf_tracking_memory_provider_t *)provider;
+
+    // umfMemoryTrackerRemove should be called before umfMemoryProviderFree
+    // to avoid a race condition. If the order would be different, other thread
+    // could allocate the memory at address `ptr` before a call to umfMemoryTrackerRemove
+    // resulting in inconsistent state.
+    if (ptr) {
+        // TODO: below method requires size but does not use it internally
```

Comment:
please either keep it w/o TODOs or remove it :) btw we could always add it later. What about CUDA? Do they use it?

Reply:
ok so I suggest to remove size param